lib/async_hooks.js

```diff
@@ -461,6 +462,19 @@ function init(asyncId, type, triggerId, resource) {
   processing_hook = false;
 }
 
+const JSProviders = new Set();
+
+function registerTypeName(type) {
+  if (JSProviders.has(type))
+    throw new errors.TypeError('ERR_ASYNC_PROVIDER_NAME', type);
```

Comment:
could do:
```js
const JSProviders = new Map();
function registerTypeName(type) {
  if (JSProviders.has(type))
    throw JSProviders.get(type);  // include the old stack somehow
  JSProviders.set(type, new errors.TypeError('ERR_ASYNC_PROVIDER_NAME', type));
}
```
Will throw weird errors, good for the AsyncResource implementer, bad for a 6-degrees-of-separation users